Avoid spawning too many idle threads

guowangy opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid spawn too many idle threads

### Documentation entry for user-facing changes
For short running queries (like ClickBench Q0), spawning too many threads is the performance bottleneck, most of the time wasted in launching threads, then waiting for all threads to finished. Based on these observations, 2 code changes introduced:
1. Only allow one to spawn threads since most of the code is serialized. This try to fix the problem that in serialized pipeline, the thread has been assigned the key task for next loop but stuck in spawning threads, the task can only be executed after all spawning done.
2. Stop spawning threads if idle number exceeds the threshold, this can help to finish the pipeline in time and avoid spawning too much unused threads.

In Xeon 8380 (2 socket, 160 vCPU in total), it can improve QPS of ClickBench Q0 by 99%, for Q6, Q7, Q41 and Q42, also have 5%~7% performance gain.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

guowangy committed:
Only allow one thread to spawn in pipeline

Only allow one to spawn threads since most of the code is serialized.
This try to fix the problem that in serialized pipeline, the thread has
been assigned the key task for next loop but stuck in spawning threads,
the task can only be executed after all spawning done.

guowangy committed:
Avoid spawn too many idle threads

This can help to finish the pipeline in time and avoid spawning too much
unused threads.

guowangy committed:
Use a better way to do try_lock

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #72920 from guowangy/ClickHouse:limit-idle-threads -->
*This is an automated comment for commit 7945c4eefdf0b94c684238d339ded2ec480b935e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72920/7945c4eefdf0b94c684238d339ded2ec480b935e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/1e6a0ab07499b529472bae448b4edc948df3ca6f/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

guowangy committed:
Fix style-check failure

guowangy committed:
Fix CI fail for the assumption of thread count

nickitat commented on src/Processors/Executors/PipelineExecutor.cpp:
```diff
@@ -379,7 +378,15 @@ void PipelineExecutor::initializeExecution(size_t num_threads, bool concurrency_
+    }
+
+    AcquiredSlotPtr slot;
+    while (tasks.shouldSpawn() && (slot = cpu_slots->tryAcquire()))
```
could we maybe use `thread_num` from l.391? would be nice to avoid taking a lock (but `shouldSpawn` currently does that)

guowangy replied:
`thread_num` is not the idle threads count. To avoid lock, maybe we can use a atoimc `idle_count` and updated if `threads_queue` size changed.

guowangy committed:
Continue fix thread count assumption

guowangy committed:
Avoid the lock in shouldSpawn()

guowangy commented on src/Processors/Executors/PipelineExecutor.cpp:
```diff
@@ -364,13 +364,12 @@ void PipelineExecutor::initializeExecution(size_t num_threads, bool concurrency_
     graph->initializeExecution(queue, async_queue);
 
-    tasks.init(num_threads, use_threads, profile_processors, trace_processors, read_progress_callback.get());
+    tasks.init(num_threads, 1, profile_processors, trace_processors, read_progress_callback.get());
```
Do you mean making the threashold adjustable by parameter?

nickitat replied:
no, i propose to start from `TOO_MANY_IDLE_THRESHOLD` instead of 1

guowangy replied:
`use_threads` is used to determine the tasks finished condition (whether all use_threads are into idle). It should reflect the number of threads spawned.  Make it to `TOO_MANY_IDLE_THRESHOLD` do not affect the performance. Furthermore, it might greater than `num_threads` (e.g `number_threads = 2`), which will make tasks finished condition unmet forever.

guowangy committed:
Fix py style check

guowangy committed:
Make test_concurrent_threads_soft_limit_defined_50  more robust

Link: https://github.com/ClickHouse/ClickHouse/pull/72920?new_mergebox=true#issuecomment-2540434416

guowangy commented:
For [test_concurrent_threads_soft_limit_defined_50](https://s3.amazonaws.com/clickhouse-test-reports/72920/4cd2501804ebddf3122a060dc93e114e04130ed9/integration_tests_flaky_check__asan_.html) failure, it is not related to this PR. But I would like to make it more robust. 

For assert, it requires `ConcurrencyControlSlotsDelayed` to equals 50, but 51 found.
```
=================================== FAILURES ===================================
______________ test_concurrent_threads_soft_limit_defined_50[1-5] ______________
....
>       assert_profile_event(
            node2,
            "test_concurrent_threads_soft_limit_2",
            "ConcurrencyControlSlotsDelayed",
            lambda x: x == 50,
        )
test_concurrent_threads_soft_limit/test.py:134: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
E       assert False
E        +  where False = <function test_concurrent_threads_soft_limit_defined_50.<locals>.<lambda> at 0x7fec6d023520>(51)
```

`ConcurrencyControlSlotsDelayed` is updated by `max - granted` and looks like `max` is assumed to `100` in CI environment, while `granted` is determined by `available(lock)` in this case.
```cpp
[[nodiscard]] SlotAllocationPtr ConcurrencyControl::allocate(SlotCount min, SlotCount max)
{
    ...
    SlotCount granted = std::max(min, std::min(max, available(lock)));
    cur_concurrency += granted;
    ProfileEvents::increment(ProfileEvents::ConcurrencyControlSlotsGranted, min);

    // Create allocation and start waiting if more slots are required
    if (granted < max)
    {
        ProfileEvents::increment(ProfileEvents::ConcurrencyControlSlotsDelayed, max - granted);
        ProfileEvents::increment(ProfileEvents::ConcurrencyControlQueriesDelayed);
        return SlotAllocationPtr(new Allocation(*this, max, granted,
            waiters.insert(cur_waiter, nullptr /* pointer is set by Allocation ctor */)));
    }
```
For `available(lock)`, here `max_concurrency` is equals to `concurrent_threads_soft_limit` (= 50 in this case), but `cur_concurrency` is vary according to system loading (e.g. background tasks). So, the `granted = available(lock) <= 50`, thus `ConcurrencyControlSlotsDelayed = (100 - granted) >= 50`.
```
SlotCount ConcurrencyControl::available(std::unique_lock<std::mutex> &) const
{
    if (cur_concurrency < max_concurrency)
        return max_concurrency - cur_concurrency;
    return 0;
}
```

guowangy commented:
The failure of [Build-binary_tidy](https://github.com/ClickHouse/ClickHouse/actions/runs/12308844500/job/34357265751?pr=72920#logs) is not related to this PR.
```
Dec 13 05:58:38 /build/src/Databases/Iceberg/RestCatalog.cpp:56:5: error: multiple declarations in a single statement reduces readability [readability-isolate-declaration,-warnings-as-errors]
Dec 13 05:58:38    56 |     std::string client_id, client_secret;
Dec 13 05:58:38       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Dec 13 05:58:38 /build/src/Databases/Iceberg/RestCatalog.cpp:626:17: error: multiple declarations in a single statement reduces readability [readability-isolate-declaration,-warnings-as-errors]
Dec 13 05:58:38   626 |                 std::string access_key_id, secret_access_key, session_token;
Dec 13 05:58:38       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Dec 13 05:58:38 268669 warnings generated.
Dec 13 05:58:38 Suppressed 269317 warnings (268667 in non-user code, 650 NOLINT).
Dec 13 05:58:38 Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Dec 13 05:58:38 2 warnings treated as errors
```

guowangy commented:
Hi @nickitat

Can you help to review?  
The CI failures about thread count assumption have been fixed. Now we have only the build-binary-tidy error and it's not related to this PR.

nickitat committed:
Merge branch 'master' into limit-idle-threads

rschu1ze commented on src/Processors/Executors/ExecutorTasks.h:
```diff
@@ -38,13 +38,18 @@ class ExecutorTasks
     /// A set of currently waiting threads.
     ThreadsQueue threads_queue;
 
+    const static size_t TOO_MANY_IDLE_THRESHOLD = 4;
```
Please add a comment how you found constant = 4. Fine to document it as "Found by rolling dice.".

rschu1ze commented on src/Processors/Executors/PipelineExecutor.h:
```diff
@@ -85,6 +85,7 @@ class PipelineExecutor
     AcquiredSlotPtr single_thread_cpu_slot; // cpu slot for single-thread mode to work using executeStep()
     std::unique_ptr<ThreadPool> pool;
     std::atomic_size_t threads = 0;
+    std::mutex spawn_lock;
```
Please use TSA annotations to document what the lock protects (... grep "TSA_" in the source code for examples).

rschu1ze commented on src/Processors/Executors/ExecutorTasks.h:
```diff
@@ -67,6 +72,8 @@ class ExecutorTasks
 
     void processAsyncTasks();
 
+    bool shouldSpawn() const { return idle_threads <= TOO_MANY_IDLE_THRESHOLD; }
```
Maybe I miss something but I find the name "idle" in "idle_threads" weird. These describe spawned threads which are doing work and they are _not_ idling, right? Why not call the variable "spawned_threads" and the constant "TOO_MANY_SPAWNED_THREADS".

guowangy replied:
`spawned_threads` should roughly equal to `use_threads`.

The "spawned" threads are not idle unless placed into `threads_queue`. This is why `idle_threads` is introduced, to dynamically tracked number of idle threads.

rschu1ze commented on src/Processors/Executors/PipelineExecutor.cpp:
```diff
@@ -364,13 +364,12 @@ void PipelineExecutor::initializeExecution(size_t num_threads, bool concurrency_
     graph->initializeExecution(queue, async_queue);
 
-    tasks.init(num_threads, use_threads, profile_processors, trace_processors, read_progress_callback.get());
+    tasks.init(num_threads, 1, profile_processors, trace_processors, read_progress_callback.get());
```
I don't know where exactly (maybe here) but please document somewhere the purpose of the optimization, i.e. why the thread count is initially 1 and incremented incrementally afterwards. Otherwise, readers of this code will need to do Git archeology.

guowangy replied:
Because one of the key conditions to determine whether tasks are finished is num of `use_threads` threads all going into idle (placed into `threads_queue`). I think this variable should somewhat reflect number of spawned threads, otherwise, we need to wait for all threads spawned even if these is no job to do. 

Setting to 1 instead of 0 is to tackle the problem in single thread. 

Will add the doc.

guowangy committed:
Merge branch 'master' into limit-idle-threads

guowangy committed:
TSA and comments update

nickitat commented:
`Integration tests (aarch64) [4/6]` - #73639
`Integration tests (asan, old analyzer) [6/6]` - #73639
`Integration tests (tsan) [6/6]` - #73639
